packages/core/src/render3/instructions/shared.ts

```diff
@@ -1663,10 +1691,70 @@ function refreshComponent(hostLView: LView, componentHostIdx: number): void {
   ngDevMode && assertEqual(isCreationMode(hostLView), false, 'Should be run in update mode');
   const componentView = getComponentLViewByIndex(componentHostIdx, hostLView);
   // Only attached components that are CheckAlways or OnPush and dirty should be refreshed
-  if (viewAttachedToChangeDetector(componentView) &&
-      componentView[FLAGS] & (LViewFlags.CheckAlways | LViewFlags.Dirty)) {
+  if (viewAttachedToChangeDetector(componentView)) {
     const tView = componentView[TVIEW];
-    refreshView(componentView, tView, tView.template, componentView[CONTEXT]);
+    if (componentView[FLAGS] & (LViewFlags.CheckAlways | LViewFlags.Dirty)) {
+      // pre-clear contains dirty in case a subtree re-adds it
+      componentView[FLAGS] &= ~(LViewFlags.ContainsDirty);
+      refreshView(componentView, tView, tView.template, componentView[CONTEXT]);
+    } else if (
+        componentView[FLAGS] & LViewFlags.ContainsDirty ||
+        componentView[CONTAINS_ALWAYS_COUNT] > 0) {
+      refreshContainsDirtyComponent(hostLView, componentHostIdx);
+    }
+  }
+}
+
+function refreshContainsDirtyComponent(hostLView: LView, componentHostIdx: number): void {
```

Comment:
```suggestion
function descendComponent(hostLView: LView, componentHostIdx: number): void {
```

1. I would like to reserve `refresh` for actually running CD, and descend to just go into, but not refresh.
2. also it is not `containsDirty` as it could also have `alwaysCount`.